### ISSUE TITLE
A corner-case where type cannot be inferred

### ISSUE THREAD

**Issue by lkct** [microsoft/pyright#9023]
(sorry I'm not sure how to write a more informative title for this case. feel free to modify :)

**Describe the bug**

As in the code below, the `reveal_type` should work in the `self_list += ...` line, but the actual behaviour is nothing for that line. Also, mouse hovering (in pyright playground) gives `Unknown` for both `another_self` and `another_self.list_int`; nevertheless, mouse hovering on `type(another_self)` correctly produces `(variable) another_self: Self@Foo`.

I have tried my best to minimise the code sample, and here are some points I want to note:
- The try-block seems useless, but if you remove it, or even remove only the `if flag1` sub-block, or only the `continue` statement, the behaviour would become normal.
- Directly assigning `self` to `another_self` instead of popping it from the list (inferred as `list[Self]`) would also make it work.
- Using either `append[...]` or `extend([...])` works, only `+=[...]` fails.
- Changing `type(another_self)` to `type(self)` also makes it work (even still using `another_self` inside `reveal_type`).

**Code or Screenshots**

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMogCmAboQIYA2A%2BvAoQFCMDGFZAzm1AGJhgBc9KEKgUkbGFVQw%2BIsTADaUgLqNhUACaFgUKpJRIJVABRtCFYABooZEGgCMMqQEoBatafMA6UeL34AvFDyNvYqgsKa2gBGNiZmllDArPYyUbwULuFuUB7AVD4BQblh2UIA7gAWSBSEOfH5cq6lamQoYDAVhCBUuVCBuQ3injgIRk6qzcIwIHBNk24I7GxZk4QAHkyECPgA8gDKAKIg4CBz88JI2klkKSvnC0t350xgKDCoAK4MT9lEpJQ0RCEIytdqdbq5bxyPzje4DApQADUgXktGBoI6XR68ScRj%2B5GoaJBbUxEPiUN8zicKnu%2BIBRIx4OxXgKMPoQA)

```python
from typing import reveal_type

class Foo:
    list_int: list[int]

    def __init__(self, arg1: int):
        self.list_int = [arg1]

    def bar(self, flag1: bool):
        self_list = [self]

        while self_list:
            another_self = self_list.pop()

            try:
                pass
            except OSError:
                if flag1:
                    pass
                continue

            reveal_type(another_self.list_int)  # Type of "another_self.list_int" is "list[int]"
            self_list += [type(another_self)(reveal_type(another_self.list_int))]  # nothing
            reveal_type(another_self.list_int)  # Type of "another_self.list_int" is "list[int]"
```

**VS Code extension or command-line**
pyright latest (1.381) as provided by the playground

<details>
<summary>A bit difference on VS Code</summary>

There might be some strange interaction with my local environment, so this can possibly be ignored, in favour of the more stable environment provided by the playground. However, it's also possible that some different bugs exist.

My local env: VS Code 1.93.1 (38c31bc7), extension: Python v2024.14.1, Pylance v2024.9.1

The code above behaves slightly differently in VS Code. The `reveal_type` still yields nothing in the `+=` line, but mouse hovering can give `(variable) list_int: list[int]`. With a bit more complication as below, I can still make hovering produce `Unknown`.

```python
from typing import reveal_type


class Foo:
    list_int: list[int]

    def __init__(self, arg1: int):
        self.list_int = [arg1]

    def bar(self, flag1: bool):
        self_list = [self]

        while self_list:
            another_self = self_list.pop()

            try:
                pass
            except OSError:
                if flag1:
                    pass
                continue

            # --- Added the below ---
            some_list = [1]

            if flag1:
                yield another_self
            # --- Added the above ---

            reveal_type(another_self.list_int)
            self_list += [type(another_self)(reveal_type(another_self.list_int)) for _ in some_list]  # Modified this line
            reveal_type(another_self.list_int)
```
</details>

**Comment by erictraut** [comment]
Thanks for the bug report and the great code sample. I was able to repro the problem and identify the root cause. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.382.
